Rewrite smtp Unittest Tests To Pytest Style

punitaojha opened the pull request:
Rewrite smtp Unittest Tests To Pytest Styletest Functions.

.coverage.rc  --> Remove homeassistant/component/smtp/notify.py from
omit section.
test_notify.py --> Convert test case to pytest style.
test.jpg, test.pdf --> Add test files.

Issue id:- #40894

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I have rewritten the unittest for smtp module to pytest style functions. As
previously, we followed the unit test style of writing test cases in our project 
and we are now using pytest style. 
Here is the link for the issue: https://github.com/home-assistant/core/issues/40894 


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests





## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #40894 
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

punitaojha committed:
Rewrite smtp unittest tests to pytest style
test functions.

.coverage.rc  --> Remove homeassistant/component/smtp/notify.py from
omit section.
test_notify.py --> Convert test case to pytest style.
test.jpg, test.pdf --> Add test files.

Issue id:- #40894

homeassistant commented:
Hi @punitaojha, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#41720).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented on tests/components/smtp/test_notify.py:
```diff
@@ -139,3 +68,83 @@ async def test_reload_notify(hass):
+    yield mailer
+
+
+html = """
```
Please use capital letters for constants.
```suggestion
HTML = """
```

punitaojha replied:
Changed to capital. will submit pull request in next commit.

MartinHjelmare commented on tests/components/smtp/test_notify.py:
```diff
@@ -139,3 +68,83 @@ async def test_reload_notify(hass):
+            </body>
+        </html>"""
+
+email_data = [
```
```suggestion
EMAIL_DATA = [
```

punitaojha replied:
Changed to capital, will submit in next commit.

MartinHjelmare commented on tests/components/smtp/test_notify.py:
```diff
@@ -139,3 +68,83 @@ async def test_reload_notify(hass):
+
+@pytest.mark.parametrize(
+    "email_data",
+    [email_data[0], email_data[1], email_data[2], email_data[3], email_data[4]],
```
```suggestion
    EMAIL_DATA,
```

punitaojha replied:
Hi @MartinHjelmare, I don't think this is correct.
we can change Line 116 to 
`[EMAIL_DATA[0], EMAIL_DATA[1], EMAIL_DATA[2], EMAIL_DATA[3], EMAIL_DATA[4]],`

because Parametrize work in a for loop format and takes it each of these data from array for checking every case of the test. 
Waiting for your opinion here.

MartinHjelmare replied:
It's the same:
```
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: EMAIL_DATA = [
   ...:     ("Test msg"),
   ...:     ("Test msg", {"images": ["tests/testing_config/media/test.jpg"]}),
   ...:     ("Test msg", {"html": "html", "images": ["tests/testing_config/media/test.jpg"]}),
   ...:     ("Test msg", {"html": "html", "images": ["test.jpg"]}),
   ...:     ("Test msg", {"html": "html", "images": ["tests/testing_config/media/test.pdf"]}),
   ...: ]

In [2]: EMAIL_DATA == [EMAIL_DATA[0], EMAIL_DATA[1], EMAIL_DATA[2], EMAIL_DATA[3], EMAIL_DATA[4]]
Out[2]: True
```

punitaojha replied:
Thanks for the explanation. Fixed this. Will submit in the next push.

MartinHjelmare commented on .coveragerc:
```diff
@@ -799,7 +799,6 @@ omit =
     homeassistant/components/smarthab/cover.py
     homeassistant/components/smarthab/light.py
     homeassistant/components/sms/*
-    homeassistant/components/smtp/notify.py
```
Coverage needs to be 95+ % per module to not lower our average. If we don't reach that, we can keep the file excluded from coverage.

punitaojha replied:
done, submitted in a new commit.

MartinHjelmare commented on tests/components/smtp/test_notify.py:
```diff
@@ -139,3 +68,83 @@ async def test_reload_notify(hass):
+email_data = [
+    ("Test msg"),
+    ("Test msg", {"images": ["tests/testing_config/media/test.jpg"]}),
+    ("Test msg", {"html": html, "images": ["tests/testing_config/media/test.jpg"]}),
+    ("Test msg", {"html": html, "images": ["test.jpg"]}),
+    ("Test msg", {"html": html, "images": ["tests/testing_config/media/test.pdf"]}),
```
```suggestion
    ("Test msg", {"html": HTML, "images": ["tests/testing_config/media/test.jpg"]}),
    ("Test msg", {"html": HTML, "images": ["test.jpg"]}),
    ("Test msg", {"html": HTML, "images": ["tests/testing_config/media/test.pdf"]}),
```

punitaojha replied:
Fixed on my system, will submit with next pull request, I make.

MartinHjelmare commented on tests/components/smtp/test_notify.py:
```diff
@@ -139,3 +68,83 @@ async def test_reload_notify(hass):
+    """Verify if we can send messages of all types correctly."""
+    sample_email = "<mock@mock>"
+    with patch("email.utils.make_msgid", return_value=sample_email):
+        if isinstance(email_data, tuple):
+            result = message.send_message(email_data[0], data=email_data[1])
+        elif isinstance(email_data, str):
+            result = message.send_message(email_data)
```
Please make the parameters consistent so we don't need if checks in the test. Set the data to `None` at minimum.
```suggestion
        result = message.send_message(email_data[0], data=email_data[1])
```

punitaojha replied:
I fixed this and will submit in next commit.

MartinHjelmare commented on tests/components/smtp/test_notify.py:
```diff
@@ -139,3 +68,83 @@ async def test_reload_notify(hass):
+        </html>"""
+
+email_data = [
+    ("Test msg"),
```
```suggestion
    ("Test msg", None),
```

punitaojha replied:
Fixed this and will submit in next commit.

MartinHjelmare commented on tests/components/smtp/test_notify.py:
```diff
@@ -139,3 +68,83 @@ async def test_reload_notify(hass):
+            "Test msg$"
+        )
+        if isinstance(email_data, str):
+            assert re.search(expected, result)
```
I would try to make parameters of this too, so we don't need to have any if checks.

punitaojha replied:
Sure let me know once we have that.

MartinHjelmare replied:
I mean that we can extend the parameters in the pytest parametrization to avoid using checks in the test.

MartinHjelmare replied:
If that's hard to do I would prefer to define multiple separate tests without pytest.parametrize which also would avoid checks inside the tests.

MartinHjelmare reviewed (changes requested):
Thanks! Try to build the parameter data so we can avoid the `if` checks in the tests.

punitaojha commented:
@MartinHjelmare Thanks for the quick review, I will make the requested changes and will reply in each thread.

punitaojha committed:
Made Fix Based On Code Review
To Pull Request -  #41720.

.coveragerc --> Add file in omit section as coverage is 75%(i.e. <95%)
test_notify.py --> Capitalize constant in file, Made parametrized data
compatible to avoid if conditions.

Code Review Submission on Pull Request #41720.

punitaojha commented:
@MartinHjelmare  I have fixed all the above issues. And submitted a new commit. Please review.

punitaojha committed:
Made Fix Based On Code Review
To Pull Request #41720.

test_notify.py --> Add different test for simple text message to avoid if conditions check in code.

punitaojha commented:
@MartinHjelmare I wrote different test cases to get rid of if condition. Can you please review it? I believe it involves all suggested changes now.

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare reviewed (changes requested):
I think our fixture media files are interfering with the media browse tests. Let's move the jpg and pdf files to another new directory `tests/testing_config/notify`.

punitaojha commented:
Thanks for the approval. Can you please share your reasons for not wanting to have if condition in the code. I know a few of them but would love to know yours.

MartinHjelmare commented:
I want to keep the tests "declarative" in spirit and clean of logic that could break. It's also easier to see what a test should do if you don't need to read if/else logic.

punitaojha committed:
Move Notify Media Files To
 Folder tests/testing_config/notify.

 test_notify.py --> Add Relevant Path For .jpg and .pdf file.
 test.jpg, test.pdf --> Move to relevant location.

 Pull request #41720

MartinHjelmare reviewed (approved):
Great!
